sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -470,3 +397,427 @@ object JoinReorderDPFilters extends PredicateHelper {
+    plan match {
+      case Some(joinPlan) =>
+        // We use the negative cost as fitness.
+        - joinPlan.planCost.card.toDouble * conf.joinReorderCardWeight -
+            joinPlan.planCost.size.toDouble * (1 - conf.joinReorderCardWeight)
+      case _ =>
+        - Double.MaxValue
+    }
+  }
+}
+
+/**
+ * A live space which has multi individuals(represented by [[Chromosome]]).
+ * The population can evolve, generation by generation. The child individual
+ * is generated by the [[Crossover]] procedure from its parents.
+ */
+object Population {
+  def apply(
+      conf: SQLConf,
+      itemsMap: Map[Int, JoinPlan],
+      conditions: Set[Expression],
+      topOutputSet: AttributeSet) : Population = {
+
+    val chromos: Seq[Chromosome] = Seq.fill(determinePopSize(conf, itemsMap.size)) {
+      Chromosome(conf, shuffle(itemsMap), conditions, topOutputSet)
+    }
+
+    Population(conf, chromos)
+  }
+
+  private def determinePopSize(conf: SQLConf, numRelations: Int): Int = {
+    val relaxFactor = conf.joinReorderGARelaxFactor
+    // The default population size:
+    // # of relations | pop size (RF=3) | pop size (RF=3.5)| pop size  (RF=4)
+    //  < 13          |   DP based      |   DP based       |   DP based
+    //    13          |   20            |   16 (13<16)     |   16
+    //    14          |   25            |   16             |   16
+    //    15          |   32            |   19             |   16
+    //    16          |   40            |   23             |   16
+    //    17          |   50            |   28             |   19
+    //    18          |   64            |   35             |   22
+    //    19          |   80            |   43             |   26
+    //    20          |   101           |   52             |   32
+    //    21          |   128           |   64             |   38
+    //    22          |   128           |   78             |   45
+    //    23          |   128           |   95             |   53
+    //    24          |   128           |   115            |   64
+    //    25          |   128           |   128            |   76
+    //    26          |   128           |   128            |   90
+    //    27          |   128           |   128            |   90
+    //    28          |   128           |   128            |   128
+    //  > 28          |   128           |   128            |   128
+    val size = math.pow(2.0, numRelations / relaxFactor)
+    val max = conf.joinReorderGAMaxPoPSize
+    val min = conf.joinReorderGAMinPoPSize
+
+    math.ceil(math.max(math.min(max, size), min)).toInt
+  }
+
+  private def shuffle(itemsMap: Map[Int, JoinPlan]) : Seq[JoinPlan] = {
+    util.Random.shuffle(itemsMap.values).toSeq
+  }
+}
+
+case class Population(conf: SQLConf, chromos: Seq[Chromosome]) extends Logging {
+
+  def evolve: Population = {
+    // Sort chromos in the population first.
+    var tempChromos = chromos.sortWith((left, right) => left.fitness > right.fitness)
```

Comment:
I know naming is hard in computer science :-) But we can definitely call it `evolvedChromos` or something, which makes more sense. Reviewers/developers can immediately realized that it stores _evolved_ chromos.

Reply:
Done.